[WIP][TensorExpr] Add a tensor expressions tutorial.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45453 [WIP][TensorExpr] Add a tensor expressions tutorial.**
* #45520 New work wihtout tutorial update



Differential Revision: [D23972887](https://our.internmc.facebook.com/intern/diff/D23972887)

Commit:
[WIP][TensorExpr] Add a tensor expressions tutorial.

[ghstack-poisoned]

Lilyjjo commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,128 @@
+    // placeholder similar to Var, but with dimensions info.
+    //
+    // Let's construct a simple load:
+    BufHandle A("A", {ExprHandle(64), ExprHandle(32)}, kInt);
```
Is a BufHandle a 'Buf expression'?

ZolotukhinM replied:
Yes, `Buf` is an `Expr` too. And similarly to how `Expr` has `ExprHandle`, `Buf` has `BufHandle`. FWIW, `Var` is also an `Expr`, and also has `VarHandle`.

"Buf" might be not the best name for this and there are chances it gets renamed soon.

Lilyjjo commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,128 @@
+    //
+    // Let's construct a simple load:
+    BufHandle A("A", {ExprHandle(64), ExprHandle(32)}, kInt);
+    Placeholder A_buf(A);
```
Is there a quick way to explain why a Placeholder is needed to wrap the BufHandle?

ZolotukhinM replied:
The primary reason is to expose a separate class in users' API while hiding internal classes (the definition of users is vague here, but you could think of several layers of details in API: the highest level of abstraction is when you use `Compute` API and never construct expressions manually). In this sense, `Tensor` and `Placeholder` are what we expose and `Buf` is what they are built upon internally.

But technically `Placeholder` doesn't add anything to what `Buf` already has.

Commit:
Update on "[WIP][TensorExpr] Add a tensor expressions tutorial."


Differential Revision: [D23972887](https://our.internmc.facebook.com/intern/diff/D23972887)

[ghstack-poisoned]
